core/arch/interp.c

```diff
@@ -3494,6 +3510,12 @@ build_bb_ilist(dcontext_t *dcontext, build_bb_t *bb)
             if (!instr_valid(bb->instr))
                 break; /* before eflags analysis! */
 
+#ifdef X86
+            if (my_dcontext != NULL && debug_register_fire_on_addr(bb->cur_pc)) {
```

Comment:
That was a bug spotted thanks to the test checking the address of the single step exception.
With a debug register, the single step exception happens before execution of the instruction at the address.

Therefore, I check bb->cur_pc to see if next instruction will fire a debug register, in which case a single step exception will be generated right after current instruction.